generators/utilities/src/main/scala/System.scala

```diff
@@ -39,7 +37,5 @@ class SystemModule[+L <: System](_outer: L) extends SubsystemModuleImp(_outer)
   with HasRTCModuleImp
   with HasExtInterruptsModuleImp
   with CanHaveMasterAXI4MemPortModuleImp
-  with CanHaveMasterAXI4MMIOPortModuleImp
```

Comment:
I don't think many external users will remove them. The way I see it, those features aren't used very often (at least here at Berkeley) and if those mixins are added they will just add 2 lines to every configuration for no reason. 

With that being said, if consensus is to have the two extra mixins, I can do that.

Reply:
I agree the additional lines to every config is verbose, but for the sake of discoverability, I think its useful to present these to external users. I also don't want a regression in functionality here, since in the future if someone does want these, it would be very annoying to dig through rocketchip and figure out how to re-add these to the system.

Reply:
I think you can make that argument with alot of other RC components. Thoughts @alonamid

Reply:
I'd delete the trait. I think the confusion created by including an AXI-related trait without producing an AXI port outweighs the pros of discoverability.

Reply:
I argue the confusion is resolved by the `WithNoMMIOPort` mixins that would be added to the configs. It should be clear that the System allows for the generation of a AXI MMIO port, but the default configs all do not use this feature.

Reply:
While I'm very pro-discoverability, I think that discoverability should be in the form of "the trait exists in a central traits file, and I can look there to find if there is something that fits my needs".  I don't think we should have stuff like `WithNoMMIOPort`

Reply:
> While I'm very pro-discoverability, I think that discoverability should be in the form of "the trait exists in a central traits file, and I can look there to find if there is something that fits my needs". I don't think we should have stuff like `WithNoMMIOPort`

Ok so there are traits and mixins. 

The `CanHaveMasterAXI4MMIOPort` trait alters the System/Subsystem, and these are distributed all over the place, so the argument that we should not include the `CanHaveMasterAXI4MMIOPort` mixin because it is defined elsewhere is flawed. Just look at all the other traits we add in to the System.

The `WithNoMMIOPort` mixin alters the config. I agree with you that mixins should be should be centrally located and commented, and they are. The `RocketConfigs.scala` file provides a well-commented list of all the mixins we use in our configs. If we were to add this mixin to the configs there, we would comment it too, of course.

Reply:
I argue that this is code that is not being used for any reason yet and most likely won't be used for a while... thus why have it in the first place. I think it is easy to re-add in the future and if someone needs it they will re-add it / PR it / ask for it.

Additionally, I think very little people knew this existed in the first place (or what it did). So I think this fine for now until someone wants it.

Reply:
Perhaps having a separate subsystem example with an AXI4 MMIO port would help alleviate the discovery concerns?

Reply:
> Perhaps having a separate subsystem example with an AXI4 MMIO port would help alleviate the discovery concerns?

But the entire reason the trait is called `CanHaveMasterAXI4MMIOPort` instead of `HasMasterAXI4MMIOPort` is so you can optionally include that port without duplicating the subsystem.

Reply:
Is this a simple change? Simple code-wise but has alot of implications on the system being created. I think John's recommendation to have a new `Top` module would match with what Chipyard is currently doing. However, I can see us going the other direction and making things like adding `GPIO`, `BlockDevice` a `CanHave***` trait and have things to disable them all. However, I think that is uglier (adds alot of extra lines to configs).